### PR TITLE
control/controlhttp: don't assume port 80 upgrade response will work

### PR DESCRIPTION
**Pull request by bradfitz**
Just because we get an HTTP upgrade response over port 80, don't
assume we'll be able to do bi-di Noise over it. There might be a MITM
corp proxy or anti-virus/firewall interfering. Do a bit more work to
validate the connection before proceeding to give up on the TLS port
443 dial.

Updates #4557 (probably fixes)

### CODE REVIEW

**Review by maisem — APPROVED** [review]
Either add a test or a TODO to add a test to make sure we don't regress here when we have a unfriendly MITM proxy

**Comment by bradfitz** [comment]
> Either add a test or a TODO to add a test to make sure we don't regress here when we have a unfriendly MITM proxy

I added a test being an unfriendly proxy and verified the test failed before and passes now.

File: control/controlhttp/http_test.go [hunk 1]
```diff
@@ -232,7 +256,14 @@ func (s *socksProxy) Start(t *testing.T) (url string) {
 	}
 	s.ln = ln
 	s.clientConnAddrs = map[string]bool{}
-	s.proxy.Logf = t.Logf
+	s.proxy.Logf = func(format string, a ...any) {
+		s.Lock()
+		defer s.Unlock()
+		if s.closed {
+			return
+		}
```
Comment (maisem) [inline control/controlhttp/http_test.go:264]: wow lol

**Review by maisem — APPROVED** [review]

File: control/controlhttp/http_test.go (hunk 1 above)
Comment (bradfitz) [reply control/controlhttp/http_test.go:264]: to fix a windows-race test flake in this code

**Pull request merged**